homeassistant/components/plugwise/climate.py

```diff
@@ -0,0 +1,310 @@
+
+Originally by https://github.com/laetificat/anna-ha which is also available
+as a development/custom-component
+"""
+
+import logging
+
+import voluptuous as vol
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.climate import (PLATFORM_SCHEMA, ClimateDevice)
+from homeassistant.components.climate.const import (CURRENT_HVAC_HEAT,
+                                                    CURRENT_HVAC_IDLE,
+                                                    HVAC_MODE_AUTO,
+                                                    HVAC_MODE_OFF,
+                                                    SUPPORT_PRESET_MODE,
+                                                    SUPPORT_TARGET_TEMPERATURE)
+from homeassistant.const import (ATTR_TEMPERATURE, CONF_HOST, CONF_NAME,
+                                 CONF_PASSWORD, CONF_PORT, CONF_USERNAME,
+                                 TEMP_CELSIUS)
+from homeassistant.exceptions import PlatformNotReady
+
+SUPPORT_FLAGS = (SUPPORT_TARGET_TEMPERATURE | SUPPORT_PRESET_MODE)
+
+_LOGGER = logging.getLogger(__name__)
+
+# Configuration directives
+CONF_MIN_TEMP = 'min_temp'
+CONF_MAX_TEMP = 'max_temp'
+
+# Default directives
+DEFAULT_NAME = 'Plugwise Thermostat'
+DEFAULT_USERNAME = 'smile'
+DEFAULT_TIMEOUT = 10
+DEFAULT_PORT = 80
+DEFAULT_ICON = 'mdi:thermometer'
+DEFAULT_MIN_TEMP = 4
+DEFAULT_MAX_TEMP = 30
+
+# HVAC modes
+ATTR_HVAC_MODES = [HVAC_MODE_AUTO, HVAC_MODE_OFF]
+
+# Read platform configuration
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    vol.Optional(CONF_USERNAME, default=DEFAULT_USERNAME): cv.string,
+    vol.Optional(CONF_MIN_TEMP, default=DEFAULT_MIN_TEMP): cv.positive_int,
+    vol.Optional(CONF_MAX_TEMP, default=DEFAULT_MAX_TEMP): cv.positive_int,
+})
+
+
+def _hvac_mode(self):
+    """Return current active hvac state."""
+    domain_objects = self._api.get_domain_objects()
+    hvac_mode = self._api.get_schema_state(domain_objects)
+    if hvac_mode is True:
+        self._state = HVAC_MODE_AUTO
+    else:
+        self._state = HVAC_MODE_OFF
+    return self._state
+
+
+def _available_schemas(self):
+    """Return schemas."""
+    domain_objects = self._api.get_domain_objects()
+    return self._api.get_schema_names(domain_objects)
+
+
+def _active_schema(self):
+    """Return active schema if present."""
+    domain_objects = self._api.get_domain_objects()
+    return self._api.get_active_schema_name(domain_objects)
+
+
+def _heating_action(self):
+    """Return current active heating state."""
+    domain_objects = self._api.get_domain_objects()
+    if self._api.get_heating_status(domain_objects) is True:
+        self._state = CURRENT_HVAC_HEAT
+    else:
+        self._state = CURRENT_HVAC_IDLE
+    return self._state
```

Comment:
This is strange. I modifies internal state of the class. Set the state variable in update method.

Reply:
In hindsight it was weird, apologies for not cleaning it up like I did now